src/app/server/JointFabricDatastore.cpp

```diff
@@ -0,0 +1,284 @@
+
+    for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+    {
+        listener->MarkNodeListChanged();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR JointFabricDatastore::SetNode(NodeId nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum state)
+{
+    size_t index = 0;
+    ReturnErrorOnFailure(IsNodeIDInDatastore(nodeId, index));
+    mNodeInformationEntries[index].commissioningStatusEntry.state = state;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR JointFabricDatastore::IsNodeIDInDatastore(NodeId nodeId, size_t & index)
+{
+    for (auto & entry : mNodeInformationEntries)
+    {
+        if (entry.nodeID == nodeId)
+        {
+            index = static_cast<size_t>(&entry - &mNodeInformationEntries[0]);
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR
+JointFabricDatastore::AddGroupKeySetEntry(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet)
+{
+    VerifyOrReturnError(IsGroupKeySetEntryPresent(groupKeySet.groupKeySetID) == false, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mGroupKeySetList.size() < kMaxNodes, CHIP_ERROR_NO_MEMORY);
+
+    mGroupKeySetList.push_back(groupKeySet);
+
+    return CHIP_NO_ERROR;
+}
+
+bool JointFabricDatastore::IsGroupKeySetEntryPresent(uint16_t groupKeySetId)
+{
+    for (auto & entry : mGroupKeySetList)
+    {
+        if (entry.groupKeySetID == groupKeySetId)
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+CHIP_ERROR JointFabricDatastore::RemoveGroupKeySetEntry(uint16_t groupKeySetId)
+{
+    for (auto it = mGroupKeySetList.begin(); it != mGroupKeySetList.end(); ++it)
+    {
+        if (it->groupKeySetID == groupKeySetId)
+        {
+            mGroupKeySetList.erase(it);
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR
+JointFabricDatastore::UpdateGroupKeySetEntry(
+    Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet)
+{
+    for (auto & entry : mGroupKeySetList)
+    {
+        if (entry.groupKeySetID == groupKeySet.groupKeySetID)
+        {
+            bool field_updated = memcmp(&entry, &groupKeySet,
+                                        sizeof(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type)) == 0;
+            entry              = groupKeySet;
+
+            if (field_updated)
+            {
+                // TODO: RefreshNodes
+            }
+
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR
+JointFabricDatastore::AddAdmin(
+    Clusters::JointFabricDatastore::Structs::DatastoreAdministratorInformationEntryStruct::Type & adminId)
+{
+    VerifyOrReturnError(IsAdminEntryPresent(adminId.nodeID) == false, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mAdminEntries.size() < kMaxNodes, CHIP_ERROR_NO_MEMORY);
```

Comment:
This check should be done against `kMaxAdminNodes`

Reply:
Addressed in the latest commit.